core/server/worker/src/main/java/alluxio/worker/DefaultAlluxioWorker.java

```diff
@@ -95,35 +88,33 @@
   private long mStartTimeMs;
 
   /**
-   * The worker ID for this worker. This is set when the block worker is initialized and may be
-   * updated by the block sync thread if the master requests re-registration.
-   */
-  private AtomicReference<Long> mWorkerId;
-
-  /**
    * Creates a new instance of {@link DefaultAlluxioWorker}.
    */
   public DefaultAlluxioWorker() {
     try {
-      mWorkerId = new AtomicReference<>();
       mStartTimeMs = System.currentTimeMillis();
-      mBlockWorker = new DefaultBlockWorker(mWorkerId);
-      mFileSystemWorker = new DefaultFileSystemWorker(mBlockWorker, mWorkerId);
-
-      mAdditionalWorkers = new ArrayList<>();
-      List<? extends Worker> workers = Lists.newArrayList(mBlockWorker, mFileSystemWorker);
-      for (WorkerFactory factory : ServerUtils.getWorkerServiceLoader()) {
-        Worker worker = factory.create(workers);
-        if (worker != null) {
-          mAdditionalWorkers.add(worker);
-        }
+      mRegistry = new Registry<>();
+      List<Callable<Void>> callables = new ArrayList<>();
+      for (final WorkerFactory factory : ServiceUtils.getWorkerServiceLoader()) {
+        callables.add(new Callable<Void>() {
+          @Override
+          public Void call() throws Exception {
+            factory.create(mRegistry);
```

Comment:
+1 for using the registry in the factory instead of the master, and only constructing masters via factory. If we make the factory a static internal class of the master, could we make all master constructors private to enforce that only the factory constructs the master?